pkg/controller/beat/common/config.go

```diff
@@ -58,44 +67,137 @@ func setOutput(cfg *settings.CanonicalConfig, client k8s.Client, associated comm
+// default config. From that point on it will be treated the same as a user provided config - it will be watched
+// and changes will be propagated to the Beat config.
+// Returns the results config, bool to indicate whether config was created from default and error.
+func getOrCreateUserConfig(params DriverParams, defaultConfig *settings.CanonicalConfig) (*settings.CanonicalConfig, bool, error) {
+	if params.Beat.Spec.Config != nil {
+		userCfg, err := settings.NewCanonicalConfigFrom(params.Beat.Spec.Config.Data)
+		if err != nil {
+			return nil, false, err
 		}
-		log.V(1).Info("Replacing ECK-managed configuration by user-provided configuration")
+		return userCfg, false, nil
 	}
 
-	return cfg.Render()
+	if params.Beat.Spec.ConfigRef == nil {
+		return nil, false, nil
+	}
+
+	secret, defaultUsed, err := getOrCreateConfigRefSecret(params.Client, params.Beat, params.Beat.Spec.ConfigRef, defaultConfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	nsn := k8s.ExtractNamespacedName(&params.Beat)
+	if err := watches.WatchUserProvidedSecrets(nsn, params.DynamicWatches(), ConfigRefWatchName(nsn), []string{secret.Name}); err != nil {
+		return nil, false, err
+	}
+
+	data, exists := secret.Data[ConfigFileName]
+	if !exists {
+		msg := fmt.Sprintf("no key %s in secret %s in namespace %s", ConfigFileName, secret.Name, params.Beat.Namespace)
+		err := fmt.Errorf(msg)
+		params.Logger.Error(err, "", "namespace", params.Beat.Namespace, "beat_name", params.Beat.Name, "secret_name", secret.Name)
+		params.Recorder().Event(&params.Beat, corev1.EventTypeWarning, events.EventReasonUnexpected, msg)
+		return nil, false, err
+	}
+	parsed, err := settings.ParseConfig(data)
+	if err != nil {
+		msg := "unable to parse configuration from secret"
+		params.Logger.Error(err, msg, "namespace", params.Beat.Namespace, "beat_name", params.Beat.Name, "secret_name", secret.Name)
+		params.Recorder().Event(&params.Beat, corev1.EventTypeWarning, events.EventReasonUnexpected, msg+": "+secret.Name)
+		return nil, false, err
+	}
+	return parsed, defaultUsed, nil
+}
+
+func getOrCreateConfigRefSecret(
+	client k8s.Client,
+	beat beatv1beta1.Beat,
+	configRef *commonv1.ConfigSource,
+	defaultConfig *settings.CanonicalConfig,
+) (*corev1.Secret, bool, error) {
+	secret := &corev1.Secret{}
+	if err := client.Get(
+		types.NamespacedName{
+			Name:      configRef.SecretName,
+			Namespace: beat.Namespace,
+		}, secret); err == nil {
+		// user provided an existing secret, just use it
+		return secret, false, nil
+	} else if !apierrors.IsNotFound(err) {
+		// can't be sure if the secret exists or not, error out to be safe
+		return nil, false, err
+	}
+
+	if defaultConfig == nil {
+		// we don't have a default config to use, no point in creating a secret
+		return nil, false, fmt.Errorf("secret %s not found and no default config to use", configRef.SecretName)
+	}
+
+	// create the secret on behalf of the user and use default config as a starting point
```

Comment:
if we move on with no defaults, I guess we can remove that feature?

Reply:
[Yes!](https://github.com/elastic/cloud-on-k8s/pull/3210#discussion_r438687518)